nexus/src/app/instance.rs

```diff
@@ -692,6 +692,15 @@ impl super::Nexus {
                     // the instance state to failed, we don't know what state
                     // the instance is in.
                     _ => {
+                        // Grab the potentially updated runtime before bumping
```

Comment:
I'm not sure but I don't think this is correct.  Suppose before all this started, the database state shows generation 1.  We fetch the instance up front and find generation 1.  We attempt to update the Sled Agent and it fails.  Previous to this PR, we might attempt to write "Failed" to the database with generation 2.  But we would find that (say) the state in the database is generation 3.  Why is it safe to clobber what's there?  Isn't it possible that that's legitimately newer than what we're trying to write?  Concretely, suppose some other request came through the same code path, did update it to "Failed", then the control plane migrated the instance to another Sled and it's running elsewhere.  In that case it'd be the wrong outcome for the database state to say "Failed" with a newer generation number.

Actually I'm not sure it's correct even to record an InstanceState of "Failed" here.  I thought the Sled Agent owns the runtime state as long as the instance is running there.  If the Instance is still running on the Sled Agent, then setting the state to "Failed" would be going split-brain (the database state would not match reality and I don't know how it would ever get resolved).

What do you think?

Reply:
> I'm not sure but I don't think this is correct. Suppose before all this started, the database state shows generation 1. We fetch the instance up front and find generation 1. We attempt to update the Sled Agent and it fails. Previous to this PR, we might attempt to write "Failed" to the database with generation 2. But we would find that (say) the state in the database is generation 3. Why is it safe to clobber what's there? Isn't it possible that that's legitimately newer than what we're trying to write? Concretely, suppose some other request came through the same code path, did update it to "Failed", then the control plane migrated the instance to another Sled and it's running elsewhere. In that case it'd be the wrong outcome for the database state to say "Failed" with a newer generation number.

Yeah, there's definitely more code to write here.

> Actually I'm not sure it's correct even to record an InstanceState of "Failed" here. I thought the Sled Agent owns the runtime state as long as the instance is running there. If the Instance is still running on the Sled Agent, then setting the state to "Failed" would be going split-brain (the database state would not match reality and I don't know how it would ever get resolved).

If the sled-agent sees a 400 from propolis, then it can bubble that up as non-fatal to the instance's state. Said another way, a 400 from propolis should mean that the instance's state hasn't changed. But if we see a 500 from propolis, we don't know what the state is, so the motivation for this block of code was to set the instance as failed.

However, looking at it now, this will set the instance state to failed for anything that isn't a 400. This may include communication errors like timeouts too. There's more code to write here too.